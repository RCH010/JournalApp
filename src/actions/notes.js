import Swal from 'sweetalert2';
import { db } from "../firebase/firebaseConfig";
import { fileUpload } from '../helpers/fileUpload';
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";


// con lo de thunk, que es lo que ayuda con el middleware para las acciones asincronas
//  tiene además del dispatch, el getState, y se puede acceder a esos, y ya del
//  get state, podemos acceder a los valores del state, o con el dispatch para
//  ejecutar una acción y se modifique algo del state
export const startNewNote = () => {
    return async (dispatch, getState) => {
        const {uid} = getState().auth;
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }

        const docRef = await db.collection(`${uid}/journal/notes`).add(newNote);
        dispatch(activeNote(docRef.id, newNote));
        dispatch(addNewNote(docRef.id, newNote));
    }
}
// Action to set a note in the state
export const activeNote = (id, note) => (
    {
        type: types.notesActive,
        payload: {
            id,
            ...note
        }
    }
)

export const addNewNote = (id, note) => ({
    type: types.notesAddNew,
    payload: {
        id,
        ...note
    }
})

export const startLoadingNotes = (uid) => {
    return async (dispatch) => {
        const notes = await loadNotes(uid);
        dispatch(setNotes(notes));
    }
}
 
export const setNotes = (notes) => ({
    type: types.noteLoad,
    payload: notes
});

// action to update notes in firestore
export const startSaveNote = (note) => {
    return async (dispatch, getState) => {
        const {uid} = getState().auth;

        
        const noteToFirestore = {...note};
        delete noteToFirestore.id;

        if(!noteToFirestore.url) { delete noteToFirestore.url; }
        if(!noteToFirestore.body) { delete noteToFirestore.body; }

        await db.doc(`/${uid}/journal/notes/${note.id}`).update(noteToFirestore)
            .then(() => {
                dispatch(refreshNote(note.id, note));
                Swal.fire('Saved', note.title, 'success');
            })
            .catch(() => {
                Swal.fire('Something went wrong :(', note.title, 'error');
            })
    }
}

// action to update the note that's just been updated
//  so in the list of notes it is updated
export const refreshNote = (id, note) => ({
    type: types.notesUpdated,
    payload: {
        id, 
        note: {
            id,
            ...note
        }
    }
})

// action that calls ./helpers/fileUpload to upload the image and returns
//  image url, then we call action startSaveNote so the db is updated
export const startUploadingPicture = (file) => {
    return async (dispatch, getState) => {
        const {active: activeNote} = getState().notes;
        // UI feedback
        Swal.fire({
            title: 'Uploading...',
            text: 'Please wait.',
            allowOutsideClick: false,
            showConfirmButton: false,
            willOpen: () => {
                Swal.showLoading();
            }
        }); 
        const fileUrl = await fileUpload(file);
        
        activeNote.url = fileUrl;
        dispatch(startSaveNote(activeNote));

        Swal.close();
    }
}

//
export const startDeleting = (noteId) => {
    return async (dispatch, getState) => {
        const {uid} = getState().auth;
        // deleted from firestore
        await db.doc(`${uid}/journal/notes/${noteId}`).delete();

        //now deleted from store
        dispatch(deleteNote(noteId));
    }
}

export const deleteNote = (noteId) => ({
    type: types.notesDelete,
    payload: noteId,
})

export const notesLogout = () => ({
    type: types.notesLogoutCleaning,
})
