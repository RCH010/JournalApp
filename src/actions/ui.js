import { types } from "../types/types"

export const setError = (errorMsg) => {
    return {
        type: types.uiSetError,
        payload: errorMsg
    };
};

export const removeError = () => {
    return {
        type: types.uiRemoveError
    };
};

export const startLoading = () => {
    return {
        type: types.uiStartLoading
    }
};

export const finishLoading = () => {
    return {
        type: types.uiFinishLoading
    };
}