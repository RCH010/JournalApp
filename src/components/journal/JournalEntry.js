import React from 'react'

export const JournalEntry = () => {
    return (
        <div className='journal__entry pointer'>
            <div 
                className='journal__entry-picture'
                style={{
                    backgroundSize: 'cover',
                    backgroundImage: 'url(https://www.vets4pets.com/siteassets/species/cat/kitten/tiny-kitten-in-sunlight.jpg?w=585&scale=down)'
                }}
            >
            </div>
            <div className='journal__entry-body'>
                <p className='journal__entry-title'>
                    Un nuevo d√≠a
                </p>
                <p className='journal__entry-content'>
                    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
                </p>
            </div>
            <div className='journal__entry-date-box'>
                <span>Monday</span>
                <h4>28</h4>
            </div>
        </div>
    )
}