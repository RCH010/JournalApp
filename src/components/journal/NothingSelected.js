import React from 'react'

export const NothingSelected = () => {
    return (
        <div className='nothing__main-content'>
            <p>
                Select something
                <br />
                ...or create an entry!
            </p>
            <i className='far fa-star fa-3x mt-1' />
        </div>
    )
}
