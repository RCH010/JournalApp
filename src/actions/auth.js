import { googleAuthProvider, firebase } from "../firebase/firebaseConfig";
import { types } from "../types/types"
import { finishLoading, startLoading } from "./ui";

//  CUANDO ES UNA TAREA ASINCRONA, SE NECESITA RETORNAR UN CALLBACK
export const startLoginWithEmailPassword = (email, password) => {
    return (dispatch) => {
        dispatch(startLoading());
        firebase.auth().signInWithEmailAndPassword(email,password)
            .then(({user}) => {
                dispatch(
                    login(user.uid, user.displayName, user.email)
                );
            })
            .catch(e => {
                console.log(e);
            })
            .finally(() => {
                dispatch(finishLoading())
            })
    }
}

export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {
        firebase.auth().createUserWithEmailAndPassword(email,password)
            .then( async({user}) => {
                await user.updateProfile({displayName: name});
                dispatch(
                    login(user.uid, user.displayName, user.email)
                )
            })
            .catch( e => {
                console.log(e);
            })
    }
}

export const startGoogleLogin = () => {
    return (dispatch) => {
        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({user}) => {
                dispatch(
                    login(user.uid, user.displayName, user.email)
                )
            });
    }
}

export const login = (uid, displayName, email) => ( {
    type: types.login,
    payload: {
        uid,
        displayName,
        email
    }
})