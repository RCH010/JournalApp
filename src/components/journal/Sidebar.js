import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { startLogout } from '../../actions/auth';
import { startNewNote } from '../../actions/notes';
import { JournalEntries } from './JournalEntries'

export const Sidebar = () => {

    const dispatch = useDispatch();
    const user = useSelector(state => state.auth)
    const handleLogout = () => {
        dispatch(startLogout());
    }

    const handleNewEntry = () => {
        console.log('do');
        dispatch(startNewNote());
    }
    return (
        <aside className='journal__sidebar'>
            <div className='journal__sidebar-navbar mt-1'>
                <h3 className='journal__sidebar-user-name'>
                    <i className='far fa-moon' />
                    <span> {user.name}</span>
                </h3>

                <button 
                    className='btn' 
                    onClick={handleLogout}
                >
                    Logout
                </button>
            </div>

            <div className='journal__new-entry' onClick={handleNewEntry}>
                <i className='far fa-calendar-plus fa-3x' />
                <p className='mt-1'>
                    New entry
                </p>
            </div>

            <JournalEntries />
        </aside>
    )
}
