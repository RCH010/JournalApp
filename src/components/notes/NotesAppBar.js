import React from 'react'
import moment from 'moment';
import { useDispatch, useSelector } from 'react-redux'
import { startSaveNote, startUploadingPicture } from '../../actions/notes'

export const NotesAppBar = () => {
    
    const dispatch = useDispatch();
    const {active} = useSelector(state => state.notes)

    const handleSave = () => {
        dispatch(startSaveNote(active));
    }

    const handlePictureUpload = () => {
        document.querySelector('#fileSelector').click(); 
    }
    const handleFileChange = (e) => {
        const file = e.target.files[0];
        
        if(file) {
            dispatch(startUploadingPicture(file));
        }
    }

    return (
        <div className='notes__app-bar'>
            <span>{moment().format('LL')}</span>
            <input 
                id='fileSelector'
                type='file'
                style={{display: 'none'}}
                onChange={handleFileChange}
            />
            <div>
                <button  
                    className='btn'
                    onClick={handlePictureUpload}
                >
                    Add Picture
                </button>
                <button 
                    className='btn'
                    onClick={handleSave}
                >
                    Save
                </button>
            </div>
        </div>
    )
}
