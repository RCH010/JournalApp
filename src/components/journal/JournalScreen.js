import React from 'react'
import { useSelector } from 'react-redux'
import { NoteScreen } from '../notes/NoteScreen'
import { NothingSelected } from './NothingSelected'
import { Sidebar } from './Sidebar'

export const JournalScreen = () => {

    const {active} = useSelector(state => state.notes)

    return (
        <div className='journal__main-content'>
            <Sidebar />
            <main>
                {
                    (active) 
                        ? (<NoteScreen />)
                        : (<NothingSelected />)
                }
            </main>
        </div>
    )
}
