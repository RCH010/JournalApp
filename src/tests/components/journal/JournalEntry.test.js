import { mount } from "enzyme"
import thunk from "redux-thunk";
import configureStore from 'redux-mock-store';
import { Provider } from "react-redux";
import { JournalEntry } from "../../../components/journal/JournalEntry";
import { activeNote } from "../../../actions/notes";


const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {};
const note = {
    title: 'Soy un gran titulo',
    body: 'Caus here I am... the one that u loved...',
    id: '12344AA',
    url: 'http://justalink.com/pic.png'
}
let store = mockStore(initState)
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={store}>
        <JournalEntry {...note}/>
    </Provider>
)

describe('Tests in <JournalEntry/>', () => {
    
    test('should match with snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    })
    
    test('should activate the note', () => {
        wrapper.find('.journal__entry').prop('onClick')();
        expect(store.dispatch).toHaveBeenCalledWith(
            activeNote(note.id, {...note})
        );
    })
    
})
