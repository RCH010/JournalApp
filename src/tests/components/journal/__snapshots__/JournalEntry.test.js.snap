// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <JournalEntry/> should match with snapshot 1`] = `
<div
  className="journal__entry pointer animate__animated animate__slideInLefts"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(http://justalink.com/pic.png)",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      Soy un gran titulo
    </p>
    <p
      className="journal__entry-content"
    >
      Caus here I am... the one that u loved...
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Wednesday
    </span>
    <h4>
      17th
    </h4>
  </div>
</div>
`;
